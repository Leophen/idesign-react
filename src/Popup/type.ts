export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom';

export interface PopupProps {
  /**
   * 包裹层类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 弹窗内容类名
   */
  portalClassName?: string;
  /**
   * 气泡样式
   */
  style?: React.CSSProperties;
  /**
   * 气泡提示内容
   */
  content?: React.ReactNode;
  /**
   * 气泡提示位置
   * @default top
   */
  placement?:
    | 'top'
    | 'left'
    | 'right'
    | 'bottom'
    | 'top-left'
    | 'top-right'
    | 'bottom-left'
    | 'bottom-right'
    | 'left-top'
    | 'left-bottom'
    | 'right-top'
    | 'right-bottom';
  /**
   * 触发气泡出现的方式
   * @default hover
   */
  trigger?: 'hover' | 'click' | 'context-menu';
  /**
   * 手动显示气泡
   */
  visible?: boolean;
  /**
   * 气泡默认显示
   * @default false
   */
  defaultVisible?: boolean;
  /**
   * 是否禁用气泡
   * @default false
   */
  disabled?: boolean;
  /**
   * 这个值变化时手动更新位置
   */
  updateLocation?: string | number | boolean;
  /**
   * 触发气泡操作时触发
   */
  onTrigger?: (visible: boolean) => void;
}

export interface PortalProps {
  className?: string;
  style?: React.CSSProperties;
  visible?: boolean;
  content?: React.ReactNode;
  placement?: placementType;
  top: number;
  left: number;
  width: number;
  height: number;
  getRef?: (ref: React.Ref<HTMLDivElement>) => void;
}
