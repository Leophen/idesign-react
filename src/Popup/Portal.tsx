import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import ReactDOM from 'react-dom';
import { placementType, PortalProps } from './type';

// 创建气泡提示容器
let popupWrapper = document.querySelector('#i-popup-wrapper')
if (!popupWrapper) {
  popupWrapper = document.createElement('div')
  popupWrapper.className = 'i-popup-wrapper'
  popupWrapper.id = 'i-popup-wrapper'
  document.body.append(popupWrapper)
}

const Portal: React.FC<PortalProps> = (props) => {
  const {
    className,
    style,
    visible = false,
    content = '',
    placement = 'top',
    getRef,
    ...tProps
  } = props

  const getLocationStyle = (placement: placementType, trigger: PortalProps, popup: PortalProps) => {
    let popupWidth = style?.width ? Number(style?.width) : popup.width
    let popupHeight = style?.height ? Number(style?.height) : popup.height
    const xMap = {
      'top': trigger.left + ((trigger.width - popupWidth) / 2),
      'top-left': trigger.left,
      'top-right': trigger.left + (trigger.width - popupWidth),
      'bottom': trigger.left + ((trigger.width - popupWidth) / 2),
      'bottom-left': trigger.left,
      'bottom-right': trigger.left + (trigger.width - popupWidth),
      'left': trigger.left - popupWidth - 16,
      'left-top': trigger.left - popupWidth - 16,
      'left-bottom': trigger.left - popupWidth - 16,
      'right': trigger.left + trigger.width + 16,
      'right-top': trigger.left + trigger.width + 16,
      'right-bottom': trigger.left + trigger.width + 16,
    }
    const yMap = {
      'top': trigger.top - popupHeight - 16,
      'top-left': trigger.top - popupHeight - 16,
      'top-right': trigger.top - popupHeight - 16,
      'bottom': trigger.top + trigger.height + 16,
      'bottom-left': trigger.top + trigger.height + 16,
      'bottom-right': trigger.top + trigger.height + 16,
      'left': trigger.top + ((trigger.height - popupHeight) / 2),
      'left-top': trigger.top,
      'left-bottom': trigger.top + (trigger.height - popupHeight),
      'right': trigger.top + ((trigger.height - popupHeight) / 2),
      'right-top': trigger.top,
      'right-bottom': trigger.top + (trigger.height - popupHeight),
    }
    const result = {
      left: xMap[placement],
      top: yMap[placement],
      ...style
    }
    return result
  }

  const popupRef: any = useRef(null)

  useEffect(() => {
    getRef?.(popupRef)
  }, [])

  // 重置方向次数限制
  const rePlaceNum = useRef(0)

  // 更新气泡方向
  const updatePlacement = (currentPlacement: placementType) => {
    if (popupRef.current) {
      // 原触发方向
      const direction = currentPlacement.split('-')[0]
      const directionWith = currentPlacement.split('-')[1] ? '-' + currentPlacement.split('-')[1] : ''
      // 窗口
      const winWidth = window.innerWidth
      const winHeight = window.innerHeight
      // 气泡
      const rect = popupRef.current.getBoundingClientRect()
      const popupWidth = rect.width
      const popupHeight = rect.height
      const popupTop = rect.top
      const popupLeft = rect.left

      let result: string = currentPlacement
      if (popupTop < winHeight && popupLeft < winWidth && rePlaceNum.current < 3) {
        if (direction === 'top' && popupTop < 0) {
          result = 'bottom' + directionWith
        }
        if (direction === 'bottom' && (winHeight - popupHeight - popupTop < 0)) {
          result = 'top' + directionWith
        }
        if (direction === 'left' && popupLeft < 0) {
          result = 'right' + directionWith
        }
        if (direction === 'right' && (winWidth - popupWidth - popupLeft < 0)) {
          result = 'left' + directionWith
        }
        rePlaceNum.current += 1
      }

      setCurrentPlacement(result)
    }
  }

  const [styles, setStyles] = useState({})
  const [currentPlacement, setCurrentPlacement] = useState<any>(placement)

  useEffect(() => {
    updatePlacement(currentPlacement)
  })

  useEffect(() => {
    const rect = popupRef.current.getBoundingClientRect()
    setStyles(getLocationStyle(currentPlacement, { ...tProps }, rect))
  }, [visible, currentPlacement, tProps.width, tProps.height, tProps.left, tProps.top])

  const PopupNode = (
    <div
      ref={popupRef}
      className={classNames(
        'i-popup',
        className
      )}
      data-popper-placement={currentPlacement}
      style={{ ...styles, width: style?.width, height: style?.height }}
    >
      <div className="i-popup__arrow" data-popper-placement={currentPlacement} />
      {typeof content === 'object' ? content : <div className="i-popup__text">{content}</div>}
    </div>
  )

  return (
    ReactDOM.createPortal(PopupNode, popupWrapper as HTMLElement)
  )
}

export default Portal;
