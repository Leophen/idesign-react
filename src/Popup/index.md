---
nav:
  title: 组件
  path: /components
group:
  title: 弹框提示组件
  order: 5
order: 6
---

# Popup 气泡提示

## 基本用法

```tsx
import React from 'react';
import { Button, Popup } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Popup content="提示内容">
          <Button>悬浮提示</Button>
        </Popup>
      </div>
    </div>
  );
};

export default App;
```

## 不同触发方式

```tsx
import React from 'react';
import { Button, Popup } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Popup content="提示内容">
          <Button>悬浮提示</Button>
        </Popup>
        <Popup content="提示内容" trigger="click">
          <Button>点击提示</Button>
        </Popup>
        <Popup content="提示内容" trigger="context-menu">
          <Button>右击提示</Button>
        </Popup>
      </div>
    </div>
  );
};

export default App;
```

## 不同触发方向

```tsx
import React from 'react'
import { Button, Popup } from 'idesign-react'

const App = () => {
  return (
    <div className="idesign-demo-block-popup">
      <Popup content="气泡提示内容 top" placement="top">
        <Button className="Top">top</Button>
      </Popup>
      <Popup content="气泡提示内容 top-left" placement="top-left">
        <Button className="TopLeft">top-left</Button>
      </Popup>
      <Popup content="气泡提示内容 top-right" placement="top-right">
        <Button className="TopRight">top-right</Button>
      </Popup>
      <Popup content="气泡提示内容 bottom" placement="bottom">
        <Button className="Bottom">bottom</Button>
      </Popup>
      <Popup content="气泡提示内容 bottom-left" placement="bottom-left">
        <Button className="BottomLeft">bottom-left</Button>
      </Popup>
      <Popup content="气泡提示内容 bottom-right" placement="bottom-right">
        <Button className="BottomRight">bottom-right</Button>
      </Popup>
      <Popup content="气泡提示内容 left" placement="left">
        <Button className="Left">left</Button>
      </Popup>
      <Popup content="气泡提示内容 left-top" placement="left-top">
        <Button className="LeftTop">left-top</Button>
      </Popup>
      <Popup content="气泡提示内容 left-bottom" placement="left-bottom">
        <Button className="LeftBottom">left-bottom</Button>
      </Popup>
      <Popup content="气泡提示内容 right" placement="right">
        <Button className="Right">right</Button>
      </Popup>
      <Popup content="气泡提示内容 right-top" placement="right-top">
        <Button className="RightTop">right-top</Button>
      </Popup>
      <Popup content="气泡提示内容 right-bottom" placement="right-bottom">
        <Button className="RightBottom">right-bottom</Button>
      </Popup>
    </div>
  )
}

export default App
```

<API src="./index.tsx"></API>
