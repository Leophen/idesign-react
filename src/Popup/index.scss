.i-popup {
  position: absolute;
  z-index: 5500;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 4px 8px;
  color: #262626;
  font-size: 14px;
  background-color: #fff;
  border: 1px solid rgb(228, 231, 237);
  border-radius: 4px;
  box-shadow: rgba(0, 0, 0, 0.12) 0px 3px 6px -4px, rgba(0, 0, 0, 0.08) 0px 6px 16px 0px,
    rgba(0, 0, 0, 0.05) 0px 9px 28px 8px;

  &[data-popper-placement^='top'],
  &[data-popper-placement^='bottom'] {
    margin: 16px 0;
  }
  &[data-popper-placement^='left'],
  &[data-popper-placement^='right'] {
    margin: 0 16px;
  }
}

.i-popup__arrow {
  position: absolute;
  z-index: 1;
  width: 8px;
  height: 8px;
  &::before {
    position: absolute;
    width: 8px;
    height: 8px;
    background: #fff;
    transform: rotate(45deg);
    content: '';
  }
  // top+
  &[data-popper-placement^='top'] {
    bottom: -4px;
    &::before {
      border-top-left-radius: 100%;
      box-shadow: rgb(220, 220, 220) -0.5px 0px 0px 0px inset,
        rgb(220, 220, 220) 0px -0.5px 0px 0px inset;
    }
  }
  &[data-popper-placement='top'] {
    left: 50%;
    margin-left: -4px;
  }
  &[data-popper-placement='top-left'] {
    left: 16px;
  }
  &[data-popper-placement='top-right'] {
    left: calc(100% - 24px);
  }
  // bottom+
  &[data-popper-placement^='bottom'] {
    top: -4px;
    &::before {
      border-bottom-right-radius: 100%;
      box-shadow: rgb(220, 220, 220) 0px 0.5px 0px 0px inset,
        rgb(220, 220, 220) 0.5px 0px 0px 0px inset;
    }
  }
  &[data-popper-placement='bottom'] {
    left: 50%;
    margin-left: -4px;
  }
  &[data-popper-placement='bottom-left'] {
    left: 16px;
  }
  &[data-popper-placement='bottom-right'] {
    left: calc(100% - 24px);
  }
  // left+
  &[data-popper-placement^='left'] {
    right: calc(-8px / 2);
    &::before {
      box-shadow: rgb(220, 220, 220) -0.5px 0px 0px 0px inset,
        rgb(220, 220, 220) 0px 0.5px 0px 0px inset;
    }
  }
  &[data-popper-placement='left'] {
    top: 50%;
    margin-top: calc(-8px / 2);
  }
  &[data-popper-placement='left-top'] {
    top: 6px;
  }
  &[data-popper-placement='left-bottom'] {
    top: calc(100% - 14px);
  }
  // right+
  &[data-popper-placement^='right'] {
    left: -4px;
    &::before {
      box-shadow: rgb(220, 220, 220) 0.5px 0px 0px 0px inset,
        rgb(220, 220, 220) 0px -0.5px 0px 0px inset;
    }
  }
  &[data-popper-placement='right'] {
    top: 50%;
    margin-top: calc(-8px / 2);
  }
  &[data-popper-placement='right-top'] {
    top: 6px;
  }
  &[data-popper-placement='right-bottom'] {
    top: calc(100% - 14px);
  }
}
