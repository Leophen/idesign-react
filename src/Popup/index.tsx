import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import ReactDOM from 'react-dom';
import Transition from '../Transition';

type placementType =
  'top' |
  'top-left' |
  'top-right' |
  'bottom' |
  'bottom-left' |
  'bottom-right' |
  'left' |
  'left-top' |
  'left-bottom' |
  'right' |
  'right-top' |
  'right-bottom'

export interface PopupProps {
  /**
   * 气泡类名
   */
  className?: string;
  /**
   * 气泡样式
   */
  style?: React.CSSProperties;
  /**
   * 气泡提示内容
   */
  content?: React.ReactNode;
  /**
   * 气泡提示位置
   * @default top
   */
  placement?:
  | 'top'
  | 'left'
  | 'right'
  | 'bottom'
  | 'top-left'
  | 'top-right'
  | 'bottom-left'
  | 'bottom-right'
  | 'left-top'
  | 'left-bottom'
  | 'right-top'
  | 'right-bottom';
  /**
   * 触发气泡出现的方式
   * @default hover
   */
  trigger?: 'hover' | 'click' | 'context-menu';
  /**
   * 手动显示气泡
   * @default false
   */
  visible?: boolean;
  /**
   * 是否禁用气泡
   * @default false
   */
  disabled?: boolean;
  /**
   * 触发气泡操作时触发
   */
  onTrigger?: (visible: boolean) => void;
}

export interface PortalProps {
  className?: string
  style?: React.CSSProperties
  visible?: boolean
  content?: React.ReactNode
  placement?: placementType
  top: number
  left: number
  width: number
  height: number
}

// 创建气泡提示容器
let popupWrapper = document.querySelector('#i-popup-wrapper')
if (!popupWrapper) {
  popupWrapper = document.createElement('div')
  popupWrapper.className = 'i-popup-wrapper'
  popupWrapper.id = 'i-popup-wrapper'
  document.body.append(popupWrapper)
}

const Portal: React.FC<PortalProps> = (props) => {
  const {
    className,
    style,
    visible = false,
    content = '',
    placement = 'top',
    ...tProps
  } = props

  const getLocationStyle = (placement: placementType, trigger: PortalProps, popup: PortalProps) => {
    let popupWidth = style?.width ? Number(style?.width) : popup.width
    let popupHeight = style?.height ? Number(style?.height) : popup.height
    const xMap = {
      'top': trigger.left + ((trigger.width - popupWidth) / 2),
      'top-left': trigger.left,
      'top-right': trigger.left + (trigger.width - popupWidth),
      'bottom': trigger.left + ((trigger.width - popupWidth) / 2),
      'bottom-left': trigger.left,
      'bottom-right': trigger.left + (trigger.width - popupWidth),
      'left': trigger.left - popupWidth - 16,
      'left-top': trigger.left - popupWidth - 16,
      'left-bottom': trigger.left - popupWidth - 16,
      'right': trigger.left + trigger.width + 16,
      'right-top': trigger.left + trigger.width + 16,
      'right-bottom': trigger.left + trigger.width + 16,
    }
    const yMap = {
      'top': trigger.top - popupHeight - 16,
      'top-left': trigger.top - popupHeight - 16,
      'top-right': trigger.top - popupHeight - 16,
      'bottom': trigger.top + trigger.height + 16,
      'bottom-left': trigger.top + trigger.height + 16,
      'bottom-right': trigger.top + trigger.height + 16,
      'left': trigger.top + ((trigger.height - popupHeight) / 2),
      'left-top': trigger.top,
      'left-bottom': trigger.top + (trigger.height - popupHeight),
      'right': trigger.top + ((trigger.height - popupHeight) / 2),
      'right-top': trigger.top,
      'right-bottom': trigger.top + (trigger.height - popupHeight),
    }
    const result = {
      left: xMap[placement],
      top: yMap[placement],
      ...style
    }
    return result
  }

  const popupRef: any = useRef(null)

  // 更新气泡方向
  const updatePlacement = (currentPlacement: placementType) => {
    if (popupRef.current) {
      // 原触发方向
      const direction = currentPlacement.split('-')[0]
      const directionWith = currentPlacement.split('-')[1] ? '-' + currentPlacement.split('-')[1] : ''
      // 窗口
      const winWidth = window.innerWidth
      const winHeight = window.innerHeight
      // 气泡
      const rect = popupRef.current.getBoundingClientRect()
      const popupWidth = rect.width
      const popupHeight = rect.height
      const popupTop = rect.top
      const popupLeft = rect.left

      let result: string = currentPlacement
      if (popupTop < winHeight && popupLeft < winWidth) {
        if (direction === 'top' && popupTop < 0) {
          result = 'bottom' + directionWith
        }
        if (direction === 'bottom' && (winHeight - popupHeight - popupTop < 0)) {
          result = 'top' + directionWith
        }
        if (direction === 'left' && popupLeft < 0) {
          result = 'right' + directionWith
        }
        if (direction === 'right' && (winWidth - popupWidth - popupLeft < 0)) {
          result = 'left' + directionWith
        }
      }

      setCurrentPlacement(result)
    }
  }

  const [styles, setStyles] = useState({})
  const [currentPlacement, setCurrentPlacement] = useState<any>(placement)

  useEffect(() => {
    updatePlacement(currentPlacement)
  })

  useEffect(() => {
    const rect = popupRef.current.getBoundingClientRect()
    setStyles(getLocationStyle(currentPlacement, { ...tProps }, rect))
  }, [visible, currentPlacement, tProps.width, tProps.height])

  const PopupNode = (
    <div
      ref={popupRef}
      className={classNames(
        'i-popup',
        className
      )}
      data-popper-placement={currentPlacement}
      style={{ ...styles, width: style?.width, height: style?.height }}
    >
      <div className="i-popup__arrow" data-popper-placement={currentPlacement} />
      {typeof content === 'object' ? content : <div className="i-popup__text">{content}</div>}
    </div>
  )

  return (
    ReactDOM.createPortal(PopupNode, popupWrapper as HTMLElement)
  )
}

const Popup: React.FC<PopupProps> = (props) => {
  const {
    children = '',
    className,
    style,
    content,
    placement = 'top',
    trigger = 'hover',
    visible = false,
    disabled = false,
    onTrigger = () => { }
  } = props;

  const [top, setTop] = useState(0)
  const [left, setLeft] = useState(0)
  const [width, setWidth] = useState(0)
  const [height, setHeight] = useState(0)

  const setTargetLocation = (target: HTMLElement) => {
    const rect = target.getBoundingClientRect()
    setTop((rect?.top || 0) + window.scrollY)
    setLeft((rect?.left || 0) + window.scrollX)
    setWidth((rect?.width || 0))
    setHeight((rect?.height || 0))
  }

  const [innerVisible, setInnerVisible] = useState(visible)

  // 触发节点是否在指定包裹层中
  const hasParent = (node: any, parent: HTMLElement | null) => {
    while (node) {
      if (node === parent) {
        return true;
      }
      node = node.parentNode
    }
    return false;
  };

  // 打开气泡通用方法
  const switchPopup = (e: React.MouseEvent, show: boolean) => {
    // 设置气泡位置
    let currentTriggerNode: any = e.target
    while (currentTriggerNode?.parentNode === (triggerNode.current as unknown as HTMLElement).children[0]) {
      currentTriggerNode = currentTriggerNode?.parentNode
    }
    if (currentTriggerNode === triggerNode.current) {
      currentTriggerNode = currentTriggerNode.children[0]
    }
    if (show) {
      e.persist();
      setTargetLocation((currentTriggerNode))
    }
    // 切换触发事件
    onTrigger?.(show)
    // 设置气泡显示隐藏
    setInnerVisible(show)
  }

  // 关闭气泡通用方法
  const closePopup = () => {
    onTrigger?.(false)
    setInnerVisible(false)
  }

  useEffect(() => {
    setInnerVisible(visible)
  }, [visible])

  // 气泡包裹层节点
  const triggerNode = useRef(null)

  // 全局监听事件，判断点击节点是否在气泡内，以确定是否关闭气泡
  const [listenClick, setListenClick] = useState(false)
  const ifClickInPopup = (e: any) => {
    const popupNode = document.querySelector('.i-popup')
    // 点击位置在气泡外
    if (!hasParent(e.target, popupNode as HTMLElement)) {
      // 点击位置既在气泡外 又在触发节点外
      if (!hasParent(e.target, triggerNode.current)) {
        closePopup()
      }
      setListenClick(false)
      window.removeEventListener('click', ifClickInPopup)
    }
  }
  useEffect(() => {
    if (listenClick) {
      window.addEventListener('click', ifClickInPopup)
      return () => window.removeEventListener('click', ifClickInPopup)
    }
  }, [listenClick])
  // 点击触发节点
  const handleClick = (e: React.MouseEvent) => {
    if (trigger === 'click') {
      switchPopup(e, !innerVisible)
      // 气泡在关闭状态下点击 则监听下一次全局点击事件
      if (!innerVisible) {
        setTimeout(() => setListenClick(true))
      }
    } return
  }

  // 判断点击和右击节点是否在气泡内
  const [listenContextMenu, setListenContextMenu] = useState(false)
  const ifHandleInPopup = (e: any) => {
    e.preventDefault();
    const popupNode = document.querySelector('.i-popup')
    if (!hasParent(e.target, popupNode as HTMLElement)) {
      closePopup()
      setListenContextMenu(false)
      window.removeEventListener('click', ifHandleInPopup)
      window.removeEventListener('contextmenu', ifHandleInPopup)
    }
  }
  useEffect(() => {
    if (listenContextMenu) {
      window.addEventListener('click', ifHandleInPopup)
      window.addEventListener('contextmenu', ifHandleInPopup)
      return () => {
        window.removeEventListener('click', ifHandleInPopup)
        window.removeEventListener('contextmenu', ifHandleInPopup)
      }
    }
  }, [listenContextMenu])
  // 右击触发节点
  const handleContextMenu = (e: React.MouseEvent) => {
    if (trigger === 'context-menu') {
      e.preventDefault();
      switchPopup(e, !innerVisible)
      if (!innerVisible) {
        setTimeout(() => setListenContextMenu(true))
      }
    } return
  }

  // 判断悬浮节点是否在气泡内
  const ifHoverInPopup = (e: any) => {
    e.preventDefault();
    const popupNode = document.querySelector('.i-popup')
    if (!hasParent(e.target, popupNode as HTMLElement)) {
      if (e.target.parentNode !== triggerNode.current) {
        closePopup()
      }
      window.removeEventListener('mouseover', ifHoverInPopup)
    }
  }

  // 悬浮触发节点
  const handleEnter = (e: React.MouseEvent) => {
    if (trigger === 'hover') {
      switchPopup(e, !innerVisible)
      setTimeout(() => {
        window.addEventListener('mouseover', ifHoverInPopup)
      })
    } return
  }

  // 触发节点宽高变化时重定位
  useEffect(() => {
    const resizeObserver = new ResizeObserver(entries => {
      setWidth((entries[0].contentRect.width || 0))
      setHeight((entries[0].contentRect.height || 0))
    });
    resizeObserver.observe((triggerNode.current as any))
    return () => resizeObserver.disconnect()
  }, [])

  return (
    <div className='i-popup__reference'>
      <section
        onClick={handleClick}
        onMouseEnter={handleEnter}
        onContextMenu={handleContextMenu}
        ref={triggerNode}
      >
        {children}
      </section>

      <Transition
        timeout={200}
        in={innerVisible && !disabled}
        animation='fade-in'
        key='i-popup'
      >
        <Portal
          style={{ ...style }}
          className={className}
          visible={innerVisible && !disabled}
          content={content}
          placement={placement}
          top={top}
          left={left}
          width={width}
          height={height}
        />
      </Transition>
    </div>
  );
};

Popup.displayName = 'Popup';

export default Popup;
