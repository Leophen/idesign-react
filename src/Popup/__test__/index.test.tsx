import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Popup from '../index';

const App = ({ children }) => (
  <div className="app">
    <header className="app-header">app header</header>
    <div className="app-body">{children}</div>
    <footer className="app-footer">footer</footer>
  </div>
)

describe('Popup 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Popup content="提示内容">
        <button>悬浮提示</button>
      </Popup>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('trigger', async () => {
    const { getByTestId } = render(
      <Popup content="提示内容" trigger="click">
        <button data-testid='trigger'>点击提示</button>
      </Popup>,
      { wrapper: App }
    );

    expect(document.querySelector('.i-popup')).not.toBeInTheDocument()

    const user = userEvent.setup()
    await user.click(getByTestId('trigger'))

    expect(document.querySelector('.i-popup')).toBeInTheDocument()
  });

  it('disabled', async () => {
    const { getByTestId } = render(
      <Popup content="提示内容" trigger="click" disabled>
        <button data-testid='trigger'>点击提示</button>
      </Popup>,
      { wrapper: App }
    );

    expect(document.querySelector('.i-popup')).not.toBeInTheDocument()

    const user = userEvent.setup()
    await user.click(getByTestId('trigger'))

    expect(document.querySelector('.i-popup')).not.toBeInTheDocument()
  });
});

