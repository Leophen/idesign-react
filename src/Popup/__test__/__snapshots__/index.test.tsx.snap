// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-popup__reference"
  >
    <button>
      悬浮提示
    </button>
  </div>
</DocumentFragment>
`;
